[WIP,RFT] ipq806x: add multicpu port

Ansuel opened the pull request:
This is based on https://github.com/openwrt/openwrt/pull/4036

This adds multicpu port to qca8k driver and dsa driver and add support for ipq806x.

More testing is required for dsa. We also add additional patch for fast age needed for port switch and multicast offload.

Ansuel committed:
linux: introduce multi-cpu dsa patch

Add support for multi-cpu dsa. This is a reworked version of the RFC patch proposed some time ago. By default the cpus are selected in a round-robin way and the command
'ip link set PORT link CPU_PORT' can be used to change the used cpu port at runtime.
A specific function port_change_cpu_port to correctly setup the port on cpu change request.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: enable secondary cpu port on every target

- Add support for multicpu in qca8k driver
- Enable secondary cpu port in every device

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: setup dsa port to the original gmac port

Set lan port to gmac2 and wan port to gmac1.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
iproute2: add support for link set

Add support for link set useful to set CPU port for dsa drivers.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
qca8k: add multicpu support

Add support for multi cpu port to qca8k dsa driver.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: add mdio improvement patch

Add dedicated ipq8064 mdio patch that improves and fix some error and disable REGCACHE that produce wrong data from the mdiobus.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: move supported device to dsa

Move supported device to dsa qca8k driver.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: standardize mac-address for asrock g10

Move mac-address apply to 02_network instead of using a preinit script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: move reorder eth to 02_network for unifi-ac-hd

Drop special preinit script and place it in the generic 02_network script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
qca8k: add experimental support for new mode

Add experimental support for:
- aging setting
- fast age
- mirror port
- multicast db
- alert DSA for global mtu

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@rbpp can you post here output of ip link and network config?

e9hack commented:
I've trouble with this multi cpu implementation. This is without the preferred cpu hack. If the startup script qca8k_set_port isn't executed and if the lan bridge is a simple bridge without vlan, lan1/port2 and lan3/port4 doesn't have connectivity. It looks like that the qca8k initialisation connects every user port to port0/eth0 and the dsa initialisation assigns lan1/port2 and lan3/port4 to port6/eth1.

e9hack commented:
It looks like that port0 must be set in the lookup control register in function qca8k_port_bridge_join() even if port6 is used.  
```
--- a/drivers/net/dsa/qca8k.c	2021-11-01 13:53:22.306046202 +0100
+++ b/drivers/net/dsa/qca8k.c	2021-11-01 16:18:32.824485003 +0100
@@ -2227,7 +2227,7 @@ qca8k_port_bridge_join(struct dsa_switch
 	int i, ret;
 
 	cpu_port = dsa_to_port(ds, port)->cpu_dp->index;
-	port_mask = BIT(cpu_port);
+	port_mask = BIT(cpu_port) | (dsa_cpu_ports(ds) & BIT(0));
 
 	for (i = 0; i < QCA8K_NUM_PORTS; i++) {
 		if (dsa_is_cpu_port(ds, i))

```

Ansuel commented:
@e9hackthe thing is that lookup with cpu port 0 means that traffic is also directed there and i suspect things just pass there... unless it's needed to receive igmp packet that currently we receive only from cpu port0.... 
Can you do one test? hardcode the igmp packet to port6 and check how the system work? (without the workaround of setting the port mask in bridge join as both 0 and 6)

e9hack commented:
> 
> 
> @e9hack the thing is that lookup with cpu port 0 means that traffic is also directed there and i suspect things just pass there... unless it's needed to receive igmp packet that currently we receive only from cpu port0.... Can you do one test? hardcode the igmp packet to port6 and check how the system work? (without the workaround of setting the port mask in bridge join as both 0 and 6)

My problem are not igmp packets. I've a raspi connect to lan1 and a T11 running debian bullseye connected to lan2. Both running headless. Both get dhcp leases, but ping or ssh connection works with lan2 only. Ping or ssh is done from a windows pc connected via wlan. The same issue is with lan3. If I'm using lan3 for my voip phone, there is now issue. I did expect issues with lan3 as interface, because the cpu port setup in qca8k_setup() looks wrong. All lan ports are connect to port0. Odd lan ports shall be connect to port6. If I change igmp from port0 to port6, I have the same behaviour. 
```
--- a/drivers/net/dsa/qca8k.c	2021-11-01 13:53:22.306046202 +0100
+++ b/drivers/net/dsa/qca8k.c	2021-11-01 19:17:43.925520935 +0100
@@ -1635,7 +1635,7 @@ qca8k_setup(struct dsa_switch *ds)
 	 * for igmp, unknown, multicast and broadcast packet
 	 */
 	ret = qca8k_write(priv, QCA8K_REG_GLOBAL_FW_CTRL1,
-			  BIT(cpu_port) << QCA8K_GLOBAL_FW_CTRL1_IGMP_DP_S |
+			  BIT(6) << QCA8K_GLOBAL_FW_CTRL1_IGMP_DP_S |
 			  dsa_cpu_ports(ds) << QCA8K_GLOBAL_FW_CTRL1_BC_DP_S |
 			  dsa_cpu_ports(ds) << QCA8K_GLOBAL_FW_CTRL1_MC_DP_S |
 			  dsa_cpu_ports(ds) << QCA8K_GLOBAL_FW_CTRL1_UC_DP_S);
``` 
I can solve my issue by adding port0 to the lookup control register of ports which using port6 as cpu. I can solve it too, if I configure a vlan bridge.

Ansuel commented:
> I can solve it too, if I configure a vlan bridge.

ow ok so the problem is only when vlan bridge is not used?

e9hack commented:
I can solve the issue

- by adding port0 to the lookup control register
- by using the script to set all lan ports to eth0
- by using the prefered cpu port hack wich links the lan ports to eth0
- by configure a vlan bridge

Ansuel commented:
I think it does work with vlan bridge as it use arl rule and ignore lookup regs... I assume this is the problem... But still it's strange that port6 only works normally.... wonder if i'm missing something in the port change function...

The preferred port is problematic as from kernel review dts should only describe feature of the device and driver should not enforce some special configuration so the cpu port stuff must be done at userspace... that is the reason we need a working port change configuration.

Can you @e9hack can you confirm that with the current code and your additional preferred port function everything work as it should?

Also i would change the current logic and just assign everything to cpu port0 and then make the user chose what to do...
Seems the correct way to do this.
(in theory the 2 conf should not change as we will use an init.d script that will change the port at userspace)

e9hack commented:
> 
> Can you @e9hack can you confirm that with the current code and your additional preferred port function everything work as it should?

It works for me, because lan1/lan2/lan3/lan4 are linked to port0 and only wan is linked to port6. I think, I don't have trouble with wan/port5, because I've configured a vlan bridge for the wan port.  

Since nobody else does report this issue, is there a difference between ethernet for ipq806x and ath79?

e9hack commented:
I did do some more tests. I did use the patches from here only and not the preferred cpu hack or link both cpu ports to every user port. The cpu ports are assigned in a round robin way to the user ports. It looks like that the main issue is the usage of different cpu ports by user ports in the same bridge. This is my test scrip and I did reboot the router:
```
#!/bin/sh /etc/rc.common

. /lib/functions/uci-defaults.sh

START=15

boot() {

	case $(board_name) in
		tplink,archer-c7-v2-2)
			logger -t qca8k_set_port "Setup usage of eth0 and eth1 for board \"$(board_name)\""
			# Restore original implementation where eth0 (port 6) was used for the lan port
			# and the eth1 (port 0) was used for the wan port
			ip link set lan1 link eth0	# connectivity
			ip link set lan2 link eth0	# connectivity
			ip link set lan3 link eth0
			ip link set lan4 link eth0

			ip link set wan link eth1	# connectivity


#			ip link set lan1 link eth1	# connectivity
#			ip link set lan2 link eth1	# connectivity
#			ip link set lan3 link eth1
#			ip link set lan4 link eth1

#			ip link set wan link eth0	# connectivity


#			ip link set lan1 link eth1	# connectivity 
#			ip link set lan2 link eth0	# no connectivity
#			ip link set lan3 link eth1
#			ip link set lan4 link eth0

#			ip link set wan link eth0	# connectivity


#			ip link set lan1 link eth0	# no connectivity
#			ip link set lan2 link eth1	# connectivity
#			ip link set lan3 link eth0
#			ip link set lan4 link eth1

#			ip link set wan link eth1	# connectivity

			;;
		*)
			logger -t qca8k_set_port "Nothing to do for board \"$(board_name)\""
			;;
	esac
}
```
Connectivity means, ping and wan access does work from my windows pc via wlan. I didn't test lan3 and lan4.

There is difference between my setup (tp-link C7 v2) and the ipq806x setup. At my setup, port0 is eth1 and port6 is eth0. In some of my messages, I did use eth0/1 or port0/port6 wrongly.

rbpp commented:
I've just done a few tests with the current PR on my R7800. Wan port @ eth0 (as per the startup script) cannot get an ip whether I use vlan or not. However, if I set wan to eth1 (same as lan ports) AND with vlan enabled, then it works. Wan @ eth1 without vlan: no deal.

Ansuel committed:
ipq806x: add additional fix for qca8k sgmii

Fixes no traffic bug with double cpu configuration in
rgmii+sgmii configuration.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
qca8k: reset mac06 bit in PAD0 on qca8k_setup

Some bootloader set the mac06 exchange and cause problem
later in qca8k_setup. Reset the bit unconditionally to
correctly configure the CPU port.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@rbpp @e9hack can you give this a spin i found a strange bug with rgmii and sgmii configuration that caused no traffic at all O.o

e9hack commented:
@Ansuel it may be another fix necessary. dsa_tree_setup_default_cpus() is called before qca8k_setup(), but qca8k_setup() links all user ports to port0.
```
--- a/drivers/net/dsa/qca8k.c	2021-11-02 19:20:09.639478624 +0100
+++ b/drivers/net/dsa/qca8k.c	2021-11-02 21:14:57.901692459 +0100
@@ -1651,10 +1651,11 @@ qca8k_setup(struct dsa_switch *ds)
 		/* Individual user ports get connected to CPU port only */
 		if (dsa_is_user_port(ds, i)) {
 			int shift = 16 * (i % 2);
+			struct dsa_port *cpu_dp = dsa_to_port(ds, i)->cpu_dp;
 
 			ret = qca8k_rmw(priv, QCA8K_PORT_LOOKUP_CTRL(i),
 					QCA8K_PORT_LOOKUP_MEMBER,
-					BIT(cpu_port));
+					cpu_dp ? BIT(cpu_dp->index) : BIT(cpu_port));
 			if (ret)
 				return ret;
 
```

rbpp commented on target/linux/generic/pending-5.10/850-07-net-dsa-qca8k-enable-mtu_enforcement_ingress.patch:
```diff
@@ -0,0 +1,31 @@
++++ b/drivers/net/dsa/qca8k.c
+@@ -1198,6 +1198,9 @@ qca8k_setup(struct dsa_switch *ds)
+ 	/* Set max number of LAGs supported */
+ 	ds->num_lag_ids = QCA8K_NUM_LAGS;
```
+@@ -1307,6 +1307,9 @@
+ 	/* Set min a max ageing value supported */
+ 	ds->ageing_time_min = 7000;
+ 	ds->ageing_time_max = 458745000;

rbpp replied:
This patch was not apllying correctly. I fixed it as such.

rbpp replied:
With this patch refresh, all ports are working as expected, wan and lans.

Ansuel replied:
Can you put output of ip link? The fix is correct

rbpp replied:
ip link output:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1502 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:40 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1502 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff
4: lan4@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff permaddr 3c:37:86:13:25:40
5: lan3@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff
6: lan2@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff permaddr 3c:37:86:13:25:40
7: lan1@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff
8: wan@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:40 brd ff:ff:ff:ff:ff:ff
16: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:3f brd ff:ff:ff:ff:ff:ff
19: ifb4eth0: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc tbf state UNKNOWN mode DEFAULT group default qlen 32
    link/ether 5e:44:c5:9b:e7:73 brd ff:ff:ff:ff:ff:ff
21: wgserver: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/none
22: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:41 brd ff:ff:ff:ff:ff:ff
23: pppoe-wan: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1492 qdisc tbf state UNKNOWN mode DEFAULT group default qlen 3
    link/ppp
24: wlan1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 3c:37:86:13:25:42 brd ff:ff:ff:ff:ff:ff
30: ifb4pppoe-wan: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 32
    link/ether 66:4b:7c:6d:6a:7f brd ff:ff:ff:ff:ff:ff
```

So far, all my testes are good but I'll report back if I find anything.

Ansuel replied:
the mtu to 1502 is suspicious... do you have by luck an old output?

rbpp replied:
How old? I can easily check from my previous dsa based firmware built from your PR #4036. I can also get a log from a vanilla master built with swconfig but that will take a bit longer. Just tell me what you need.

Ansuel replied:
nha don't waste time if you already have a log from #4036 it would be good... but don't even try flashing just to get that log. I will investigate if the mtu is correct.

rbpp replied:
In the #4036 PR discussion, there are a number of ip link outputs posted there and they all show mtu 1502.

e9hack commented:
It looks good. I get connectivity for all 4 tests from above. At test three and four, one connection needs more time till ping does work. That's the 'ip link' output after the last test:
```
root@:~# ip link | grep mtu
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1502 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1502 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
4: wan@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-wan state UP mode DEFAULT group default qlen 1000
5: lan1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
6: lan2@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
7: lan3@eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue master br-lan state LOWERLAYERDOWN mode DEFAULT group default qlen 1000
8: lan4@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
11: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
12: br-wan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
13: br-wan.7@br-wan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
14: br-wan.2@br-wan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
15: br-tor: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
16: br-guest1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
17: br-guest2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
18: br-IoT: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
19: wg0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
20: pppoe-wan: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1492 qdisc hfsc state UNKNOWN mode DEFAULT group default qlen 3
21: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
22: wlan1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
23: wlan1-1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
24: wlan1-2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-guest1 state UP mode DEFAULT group default qlen 1000
25: wlan1-3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-guest2 state UP mode DEFAULT group default qlen 1000
26: wlan1-4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-IoT state UP mode DEFAULT group default qlen 1000
27: tun1: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN mode DEFAULT group default qlen 500
28: wlan1-5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-tor state UP mode DEFAULT group default qlen 1000
29: wlan1-6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
30: wlan1-7: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
31: ifb0: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc hfsc state UNKNOWN mode DEFAULT group default qlen 32
32: wlan0-1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
33: wlan0-2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-tor state UP mode DEFAULT group default qlen 1000
```

Ansuel commented:
@e9hack that's very nice to read. the latency is probably related to reconfiguring and make the arl rule expire (internal switch stuff i assume)

Another step would be change the logic and set everything to the first cpu port and then leave userspace decide what to use. (so remove the cpu port mixed code)

About the cpu port setup in theory it should not cause any problem as i added to the code that on port setup the port_change_cpu_port is called to make sure the correct lookup regs are set. In theory should remove any problem caused by that.

e9hack commented:
> 
> About the cpu port setup in theory it should not cause any problem as i added to the code that on port setup the port_change_cpu_port is called to make sure the correct lookup regs are set. In theory should remove any problem caused by that.

If 'ip link set ...' is not executed, because nothing is changed from user mode, port_change_cpu_port() will be not execute.

Ansuel commented:
Oh wow... i never pushed that version..........

Ansuel commented:
@e9hack did you tried doing some "on fly" test. Like 
- setting vlan to one port. 
- change cpu port
- check if vlan is still there?

e9hack commented:
> 
> 
> @e9hack did you tried doing some "on fly" test. Like

I did only a few "on the fly" tests yet, because this will not be my use case.

Changing ethx on lan4 (this is my voip phone) works immediately. Getting a phone call is possible. 1 or 2 minutes later, the phone lost the connection to the provider. Re-connecting of the cable doesn't help. Changing the ethx back does help. I didn't try a re-boot of the phone. Changing ethx on lan1, connectivity is lost. Re-connecting the cable doesn't help. Connecting the cable to lan3 or changing ethx back does help.

Ansuel commented:
mhh i would test the new version with the port change

e9hack commented:
If I change ethx on wan. Internet connection isn't lost. Vlans are retained.

Maybe I have to repeat the tests, because I'm using this additional patch to reset the port vlan id's to its default value:
```
--- a/drivers/net/dsa/qca8k.c	2021-10-22 08:20:40.630047580 +0200
+++ b/drivers/net/dsa/qca8k.c	2021-10-22 08:41:18.972965755 +0200
@@ -2548,13 +2542,50 @@ qca8k_port_vlan_del(struct dsa_switch *d
 		    const struct switchdev_obj_port_vlan *vlan)
 {
 	struct qca8k_priv *priv = ds->priv;
-	int ret = 0;
+	int tmp, ret = 0;
+	u32 reg, val;
 	u16 vid;
 
-	for (vid = vlan->vid_begin; vid <= vlan->vid_end && !ret; ++vid)
-		ret = qca8k_vlan_del(priv, port, vid);
+	for (vid = vlan->vid_begin; vid <= vlan->vid_end; ++vid) {
+		tmp = qca8k_vlan_del(priv, port, vid);
+		if (tmp < 0)
+			ret = tmp;
+	}
 
-	if (ret)
+	reg = QCA8K_EGRESS_VLAN(port);
+	tmp = qca8k_read(priv, reg, &val);
+	if (tmp < 0)
+		ret = tmp;
+	else {
+		int shift = 16 * (port % 2);
+		if (((val >> shift) & 0xfff) == vlan->vid_end) {
+			tmp = qca8k_rmw(priv, reg, 0xfff << shift,
+					QCA8K_PORT_VID_DEF << shift);
+			if (tmp < 0)
+				ret = tmp;
+		}
+	}
+
+	reg = QCA8K_REG_PORT_VLAN_CTRL0(port);
+	tmp = qca8k_read(priv, reg, &val);
+	if (tmp < 0)
+		ret = tmp;
+	else {
+		if ((val & QCA8K_PORT_VLAN_CVID(0xfff)) == QCA8K_PORT_VLAN_CVID(vlan->vid_end)) {
+			tmp = qca8k_rmw(priv, reg, QCA8K_PORT_VLAN_CVID(0xfff),
+					QCA8K_PORT_VLAN_CVID(QCA8K_PORT_VID_DEF));
+			if (tmp < 0)
+				ret = tmp;
+		}
+		if ((val & QCA8K_PORT_VLAN_SVID(0xfff)) == QCA8K_PORT_VLAN_SVID(vlan->vid_end)) {
+			tmp = qca8k_rmw(priv, reg, QCA8K_PORT_VLAN_SVID(0xfff),
+					QCA8K_PORT_VLAN_SVID(QCA8K_PORT_VID_DEF));
+			if (tmp < 0)
+				ret = tmp;
+		}
+	}
+
+	if (ret < 0)
 		dev_err(priv->dev, "Failed to delete VLAN from port %d (%d)", port, ret);
 
 	return ret;
```

Ansuel commented:
The fact that the port malfunction after 1-2 minutes seems to be related to some rules that expire after some time and are not renewed

e9hack commented:
> 
> 
> The fact that the port malfunction after 1-2 minutes seems to be related to some rules that expire after some time and are not renewed

I did change it back,  the phone did register at the provider and getting calls was possible again. Since I did change ethx for wan, the phone lost again the connection to the provider, not immediately, but I don't now after which time.

e9hack commented:
> 
> 
> It looks good. I get connectivity for all 4 tests from above. At test three and four, one connection needs more time till ping does work. That's the 'ip link' output after the _(last)_ **third** test:

It was to early to say connectivity is available. For test setup three and four (lan ports connected to both cpu ports), ping does work for a connection over the lan port which is connected to port6, but ssh will be terminate very quickly with 'Timeout, server xxxx not responding'. To get continuously traffic, I let run htop.

e9hack commented on target/linux/generic/hack-5.10/780-1-net-dsa-allow_for_multiple_CPU_ports.patch:
```diff
@@ -0,0 +1,219 @@
++
++			if (dp->ds->ops->port_change_cpu_port)
++				dp->ds->ops->port_change_cpu_port(dp->ds, dp->index, dp->cpu_dp);
++		}
```
Calling port_change_cpu_port() doesn't help, because this settings will be overwritten by qca8k_setup(), which is called after dsa_tree_setup_default_cpus().

Ansuel commented:
@rbpp @e9hack any update?

We need to understand what works and what doesn't work to think of a cause.
(lost track of this as i'm busy with leds patch)

rbpp commented:
> @rbpp @e9hack any update?

I've been using it on my R7800 and my C2600 and no problems whatsoever.

Ansuel commented:
@rbpp can you post your configuration so we can check the feature tested?

rbpp commented:
> @rbpp can you post your configuration so we can check the feature tested?

Sure. This is the relevant part of my network configuration (I don't use any VLANs):

```
config device
	option name 'br-lan'
	option type 'bridge'
	list ports 'lan1'
	list ports 'lan2'
	list ports 'lan3'
	list ports 'lan4'

config interface 'lan'
	option proto 'static'
	option ipaddr '192.168.1.1'
	list ip6class 'wan_6'
	option netmask '255.255.255.0'
	option ip6assign '64'
	option device 'br-lan'

config interface 'wan'
	option proto 'pppoe'
	option username 'xxxxxxx'
	option password 'xxxxx'
	option peerdns '0'
	option ipv6 'auto'
	option device 'wan'

config interface 'wan6'
	option proto 'dhcpv6'
	option reqaddress 'try'
	option peerdns '0'
	option reqprefix 'auto'
	option device 'wan'
```

e9hack commented:
For testing, I'm using a tplink archer C7 v2. It is ath79 based and not ipq806x. 

For lan ports, I'm using a simple bridge without vlan's. If I don't change the link order, I get the following:
```
port1/wan  <--> port0/eth1
port2/lan1 <--> port6/eth0
port3/lan2 <--> port0/eth1
port4/lan3 <--> port6/eth0
port5/lan4 <--> port0/eth1
```
This is part of my network config:
```
config interface 'lan'
	option proto 'static'
	option device 'br-lan'

config interface 'wan'
	option proto 'pppoe'
	option device 'br-wan.7'

config interface 'modem'
	option proto 'static'
	option device 'br-wan.2'

config interface 'phone'
	option proto 'static'
	option device 'lan4'

config device
	option type 'bridge'
	list ports 'tap0'
	list ports 'lan1'
	list ports 'lan2'
	list ports 'lan3'

config device
	option type 'bridge'
	list ports 'wan'

config bridge-vlan
	option device 'br-wan'
	option vlan '2'
	list ports 'wan:u'

config bridge-vlan
	option device 'br-wan'
	option vlan '7'
	list ports 'wan:t*'
```
To the lan brigde, there are wlan interfaces join too.

The connection from wlan to lan2 works without any issue. The connection from lan4 to wan works too. It is a voip phone, which is able to receive calls. The connection from wlan to lan1/lan3 doesn't really work. Ping is possible without any losses. A ssh connection will be disconnect after some time. If there is more traffic, it will be disconnected earlier. It looks like dhcp and dns are disturbed too. If I do a dhcp re-new at the client on lan1/lan3, it will be disconnect and it needs a very long time till ping works again. If I try a 'git pull', mostly I get a name resolution error. If I execute nslookup, I get an ip.

It is no issue, if lan1/lan2/lan3 are linked to the same cpu port, but I did do more tests, with lan1/lan2/lan3 linked to port0.

With an earlier state of the multi cpu patch, there was the same issue, but it was no issue, if the lan bridge was configured with vlans. I didn't do this tests with vlans with the current patch for ipq806x.

e9hack commented on target/linux/ipq806x/patches-5.10/096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch:
```diff
@@ -0,0 +1,119 @@
++
+ 		/* CPU port gets connected to all user ports of the switch */
+ 		if (dsa_is_cpu_port(ds, i)) {
+ 			ret = qca8k_rmw(priv, QCA8K_PORT_LOOKUP_CTRL(i),
```
This looks wrong:
```
@@ -1672,6 +1672,7 @@ qca8k_setup(struct dsa_switch *ds)
                /* Individual user ports get connected to CPU port only */
                if (dsa_is_user_port(ds, i)) {
                        int shift = 16 * (i % 2);
+                       cpu_port = dp->cpu_dp->index;

                        ret = qca8k_rmw(priv, QCA8K_PORT_LOOKUP_CTRL(i),
                                        QCA8K_PORT_LOOKUP_MEMBER,
```

e9hack commented on target/linux/ipq806x/patches-5.10/096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch:
```diff
@@ -0,0 +1,119 @@
+ 		if (dsa_is_user_port(ds, i)) {
++			cpu_port = dp->cpu_dp->index;
++
+ 			ret = qca8k_rmw(priv, QCA8K_PORT_LOOKUP_CTRL(i),
```
This looks wrong:
                if (dsa_is_user_port(ds, i)) {
                        int shift = 16 * (i % 2);
+                       cpu_port = dp->cpu_dp->index;

Ansuel replied:
why ?

castiel652 replied:
Patch doesn't apply

e9hack replied:
I get an error an shall fix the patch. What is the sequnce to apply patches? I copy this patch to generic/hack-5.10 and give it number 996-34.

rbpp replied:
The patch is not applying (hunk #3). If I understood @e9hack's comment correctly, the line with `int shift = 16 * (i % 2);` is required as the variable is used further down in the code.

However, if I correct the patch as per @e9hack's comment, the PR builds ok, the router works but I can't change ports with ip link set.

rbpp commented:
> For testing, I'm using a tplink archer C7 v2. It is ath79 based and not ipq806x.

@e9hack : I also tested this PR with a C7 as dumb AP (no VLANs) and it works alright with this PR. When I built this PR for my C7, I also applied the patch "096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch" by copying it into target/linux/ath79/patches-5.10 directory. Did you do the same?

e9hack commented:
> 
> 
> > For testing, I'm using a tplink archer C7 v2. It is ath79 based and not ipq806x.
> 
> @e9hack : I also tested this PR with a C7 as dumb AP (no VLANs) and it works alright with this PR. When I built this PR for my C7, I also applied the patch "096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch" by copying it into target/linux/ath79/patches-5.10 directory. Did you do the same?

Yes I did do it, but I did copy it to generic/hack-5.10 with number 996-34 to apply it after all other qca8k related patches. Did you add the script qca8k_set_port? If I change the linking between user and cpu ports, that all user ports are linked to the same cpu port, it does work. I did also add PR kernel: add PR 6542 "leds support for qca8k", but it shall not influence the switch behaviour.

rbpp commented:
> Yes I did do it, but I did copy it to generic/hack-5.10 with number 996-34 to apply it after all other qca8k related patches. Did you add the script qca8k_set_port? If I change the linking between user and cpu ports, that all user ports are linked to the same cpu port, it does work. I did also add PR kernel: add PR 6542 "leds support for qca8k", but it shall not influence the switch behaviour.

I'm not sure in what order these patches are applied. In generic/pending-5.10, there are a few qca8k patches that I understand should be applied before the multi cpu support one. So, if the patches in generic/hack-5.10 are applied before the ones in generic/pending-5.10, could there be a problem? I'm not sure... that's why I put the mult cpu support patch into the ath79 directory.

With regards to the qca8k_set_port script, I included it into /etc/rc.local.

pkgadd commented:
I'm currently failing to apply/ build this.

`target/linux/generic/backport-5.10/808-v5.16-net-dsa-qca8k-make-sure-PAD0-MAC06-exchange-is-disab.patch` conflicts with `target/linux/generic/backport-5.10/750-v5.16-net-dsa-qca8k-make-sure-pad0-mac06-exchange-is-disabled.patch`, removing `808-v5.16-net-dsa-qca8k-make-sure-PAD0-MAC06-exchange-is-disab.patch` gets me further, before failing at:

```
Applying target/linux/ipq806x/patches-5.10/096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch using plaintext: 
patching file drivers/net/dsa/qca8k.c
Hunk #1 succeeded at 1183 (offset 31 lines).
Hunk #2 succeeded at 1193 (offset 31 lines).
Hunk #3 FAILED at 1174.
Hunk #4 succeeded at 2203 (offset 217 lines).
Hunk #5 succeeded at 2278 (offset 224 lines).
1 out of 5 hunks FAILED -- saving rejects to file drivers/net/dsa/qca8k.c.rej
patching file drivers/net/dsa/qca8k.h
Hunk #1 succeeded at 174 (offset 5 lines).
Patch failed!  Please fix target/linux/ipq806x/patches-5.10/096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch!
```

e9hack commented:
@pkgadd, the third hunk is broken in 096-34-drivers-net-dsa-qca8k-add-support-for-multi-cpu-supp.patch:
```
@@ -1172,6 +1172,7 @@ qca8k_setup(struct dsa_switch *ds)
 		/* Individual user ports get connected to CPU port only */
 		if (dsa_is_user_port(ds, i)) {
 			int shift = 16 * (i % 2);
+			cpu_port = dp->cpu_dp->index;
 
 			ret = qca8k_rmw(priv, QCA8K_PORT_LOOKUP_CTRL(i),
 					QCA8K_PORT_LOOKUP_MEMBER,
```

Ansuel commented:
Anyway tomorrow will update this pr with the 5.15 and experimental pr as new depends. The idea is to test the new assisted learning on cpu port with multicpu port and check how this works.

olimpiumarius commented:
I use the latest version of OpenWrt from SNAPSHOT (r18474-87b6e10c71) on wrt1900acs v2 but i can't get past 890mbps on the wire side.
Multicpu port can be found in the latest updates?

dangowrt commented:
Apart from the new assisted learning feature which wasn't yet backported to 5.10 I've got this working on MediaTek MT7622 and MT7623 BananaPi boards, using patches from @LGA1150 and @frank-w as well:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=shortlog;h=refs/heads/bpi-multicpu

Is anyone working on submitting support for multiple CPU ports upstream? Or at least push this in official OpenWrt any time soon?

Ansuel commented:
@dangowrt multicpu upstream is still far (dsa team is working on it but we are still very far from a real patchset)
On the other hand merging this in the hack dir is a possibility mediatek marvell and qca8k would benefits from this.

dangowrt commented:
@Ansuel In the dev meeting yesterday we discussed the situation of multi-CPU-port support in DSA and decided to move forward and add it to `hack` like you suggested.
The best would be if you would send a series to the openwrt-devel mailing list adding the needed patches to `target/linux/generic/hack-5.10`, as in that way it will also get another round of attention and review on the list, potentially reaching those who are not actively following everything on Github. If you don't have time to prepare the series, I would do so and send it only to you first for pre-review and then to the list once you ACKed.
Let me know if any of that sounds agreeable to you.

Ansuel commented:
@dangowrt if this is not urgent tomorrow I should have time to clean this and send it to the mailing list.

dangowrt commented:
Not super urgent, but would be nice to have at least the generic part in master by end of next week or so. In that we at least the infrastructure is in place before we'll branch-off for the 22.xx release (which is going to happen soon now that all targets are on Linux 5.10, probably again in early February).

Ansuel commented:
@dangowrt i have bad new about this... upstream chose a totally different path for this implementation... Should we still works on this and then drop? i think only mvebu would benefits from this
ipq806x is still not ready for qca8k (perf regression)

dangowrt commented:
@Ansuel :frowning: doesn't make sense to continue that road then. any chance to get more than one CPU port working with the upstream approach?

frank-w commented:
@Ansuel can you give more info on the "new" mainline way?

Ansuel commented:
The idea is to create a LAG between the 2 CPU port and DSA will use this special interface as CPU port. They are preparing everything in DSA core for this change. Afaik Marvell switch should support this so it's really experiment and backport shitload of patch.

frank-w commented:
Ah right,found Patchset and discussion, have not yet looked if it already applied or If any optinion is against it.

https://www.spinics.net/lists/netdev/msg822508.html

I guess all dsa drivers (in my case the mt7530) amd maybe the thernet drivers (mtk_eth_soc) meed to be reworked to setup the LAG on both ends, right?

Leo-PL commented:
I have one question, possibly answered on LKML: what will determine hashing policy for that LAG? Assuming all switches can actually support this, that is.

Ansuel commented:
It should just support dsa LAG ops in theory (If supported)
Alternative way will be to bind a specific port to a specific CPU.

Il Gio 11 Ago 2022, 21:10 Frank Wunderlich ***@***.***> ha
scritto:

> Ah right,found Patchset and discussion, have not yet looked if it already
> applied or If any optinion is against it.
>
> https://www.spinics.net/lists/netdev/msg822508.html
>
> I guess all dsa drivers (in my case the mt7530) amd maybe the thernet
> drivers (mtk_eth_soc) meed to be reworked to setup the LAG on both ends,
> right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4717#issuecomment-1212383371>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQTAMPVPEGHZ6KE5BOLVYVFZRANCNFSM5GS3PIPQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Ansuel commented:
Handled here https://github.com/openwrt/openwrt/pull/4036 with changes to reflect new upstream implementation.
